### PR TITLE
Do not start a thread for an unacceptable request

### LINKED ISSUE #903: An useless thread is created on each request in the general category
The commit 01e2648e93e718f84c77aed16326a14f2b498a2e introduces a regression : 
* before this commit, the request function was calls then if an url was requested a thread was created
* with this commit a thread is created for each engine 

It's not an issue except for one engine in the general category: currency.

Most of the time the currency engine doesn't send a request, but a thread is still created (which take time)

The call to the request function in the main thread is also a problem since some engine sends an HTTP request ([wolframalpha](https://github.com/asciimoo/searx/blob/master/searx/engines/wolframalpha_noapi.py#L66) for example).

A better solution would be a function can_accept defined inside the engine ? (similar to the PR #797) 

Edit : the PR #853 could be implemented as an engine too.

### PR DESCRIPTION
**Pull request by kvch**
## What does this PR do?

This PR is the updated version of #992 created by @dalf . From now on engines can implement the function `is_accepted` to determine if the request submitted by the user can be executed. If not, the engine is never called.

## Why is this change important?

We can get rid of unnecessary threads which we know in advance won't retrieve any results.

## Related issues

Closes #992 
Closes #903

### CODE REVIEW

File: searx/engines/translated.py [hunk 1]
```diff
@@ -58,7 +63,7 @@ def response(resp):
         'title': '[{0}-{1}] {2}'.format(
             resp.search_params['from_lang'][1],
             resp.search_params['to_lang'][1],
-            resp.search_params['query']),
+            resp.search_params['query'])
```
Comment (dalf) [inline searx/engines/translated.py:66]: Without the coma, there is a syntax error:
https://travis-ci.org/github/asciimoo/searx/jobs/721829981#L938

**Comment by asciimoo** [comment]
Wouldn't it be easier to just leave the url blank in the `request` function and check it from the caller?

Minor note: it would be good to add documentation about this feature.

**Comment by dalf** [comment]
It is too late in the ```request``` function: the thread has already been started.

The callstack:
* Search.search <-- this PR checks if the engine can accept the request in this method.
* search_multiple_requests <--- here one thread is started per engine request
* search_one_request_safe
* search_one_http_request_safe
* [search_one_http_request](https://github.com/asciimoo/searx/blob/6f9aa0e2581815189db06d3e7815f1e11b9a4f15/searx/search.py#L84-L101)
* request <-- if ```request_params['url'] is None```, then there is no request, but a thread has been started for nothing.

**Comment by asciimoo** [comment]
> It is too late in the request function: the thread has already been started.

True. Then my question is: is it worth to introduce a new function just to be able to avoid starting a thread? Starting a few threads isn't that resource consuming tasks.

**Comment by dalf** [comment]
Does it worth a new function ? It requires some benchmark (on high end hardware it will be nothing, on a small SBC, perhaps it is interesting).

Some brainstorming most probably outside the scope of this PR, but still related: we can imagine that these syntaxes are generic to all currency and translation engines:
    * ```10 eur in usb```
    * ```en-it house```

In this case:
* ```is_accepted``` would be generic (one per syntax)
* the ```currency_convert``` engine declares itself as a currency convert engine
* the ```dictzone``` and ```translated``` engines declare themself as translation engines.
* the ```request``` method expect some well define parameters (currency converter --> from / to / amount)
* we can display this information to the user.

[EDIT] in the new architecture, the ```request``` and ```response``` functions will be async, and no more thread.
[EDIT2] possible implementation : https://github.com/dalf/searx/tree/processors/searx/search/processors

**Comment by kvch** [comment]
Closing this PR after an offline conversation with @asciimoo. We agreed that starting a new thread is not a big issue. The search service never gets the invalid request, so we are not putting unnecessary load on anyone. Besides @dalf is refactoring the request and respons mechanism, so threads will be gone.

**Pull request closed without merging**